src/binder/binder_util.cpp

```diff
@@ -66,9 +66,13 @@ void BinderUtil::CheckAndTryPromoteType(const common::ManagedPointer<parser::Con
         // TODO(Matt): see issue #977
         switch (desired_type) {
           case type::TypeId::DATE: {
-            auto parsed_date = execution::sql::Date::FromString(str_view);
-            value->SetValue(type::TypeId::DATE, execution::sql::DateVal(parsed_date));
-            break;
+            try {
+              auto parsed_date = execution::sql::Date::FromString(str_view);
+              value->SetValue(type::TypeId::DATE, execution::sql::DateVal(parsed_date));
+              break;
+            } catch (ConversionException &exception) {
+              throw BINDER_EXCEPTION(exception.what(), common::ErrorCode::ERRCODE_INVALID_DATETIME_FORMAT);
```

Comment:
Thanks for the PR!

1. I tried a bunch of queries and it seems that they generally prefer error 22008 ERRCODE_DATETIME_FIELD_OVERFLOW for whatever reason, could you switch to that error code?

```
postgres=# create table types_test (a date);
CREATE TABLE
postgres=# select * from types_test  where a < '0000-01-01';
ERROR:  date/time field value out of range: "0000-01-01"
LINE 1: select * from types_test  where a < '0000-01-01';
                                            ^
postgres=# \errverbose
ERROR:  22008: date/time field value out of range: "0000-01-01"
LINE 1: select * from types_test  where a < '0000-01-01';
                                            ^
LOCATION:  DateTimeParseError, datetime.c:3731
postgres=# drop table types_test;
DROP TABLE

postgres=# create table types_test (a timestamp);
CREATE TABLE
postgres=# select * from types_test  where a < '0000-01-01';
ERROR:  date/time field value out of range: "0000-01-01"
LINE 1: select * from types_test  where a < '0000-01-01';
                                            ^
postgres=# \errverbose
ERROR:  22008: date/time field value out of range: "0000-01-01"
LINE 1: select * from types_test  where a < '0000-01-01';
                                            ^
LOCATION:  DateTimeParseError, datetime.c:3731
```

2. Could you add a similar thing for timestamps right below? I believe the timestamp conversion code may also throw ConversionError.

Thanks!

Reply:
Thanks for the review @lmwnshn!

I modified it to use error code 22008 and added another try-catch for timestamps.

With regards to the error code, it's dependent on the input
```
postgres=# create table types_test (a date);
CREATE TABLE
postgres=# select * from types_test where a < '0000-01-01';
ERROR:  date/time field value out of range: "0000-01-01"
LINE 1: select * from types_test where a < '0000-01-01';
                                           ^
postgres=# \errverbose
ERROR:  22008: date/time field value out of range: "0000-01-01"
LINE 1: select * from types_test where a < '0000-01-01';
                                           ^
LOCATION:  DateTimeParseError, datetime.c:3731
postgres=# select * from types_test where a < 'dsijfdi';
ERROR:  invalid input syntax for type date: "dsijfdi"
LINE 1: select * from types_test where a < 'dsijfdi';
                                           ^
postgres=# \errverbose
ERROR:  22007: invalid input syntax for type date: "dsijfdi"
LINE 1: select * from types_test where a < 'dsijfdi';
                                           ^
LOCATION:  DateTimeParseError, datetime.c:3758
```

I believe we'd have to parse the input to get the correct error code, but I think this is out of scope for this PR. 

I actually think using error code 22007 as a catch-all would be the best option here until we can differentiate between them. I'm happy to switch this back if you agree.

Reply:
That sounds reasonable to me. Feel free to switch it back, could you please add a comment alongside it (basically what you already wrote)? Thanks!

Reply:
Great, I just updated it!